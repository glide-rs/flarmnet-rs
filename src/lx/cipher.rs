use std::io::Read;

pub fn encrypt(s: &[u8]) -> Vec<u8> {
    s.iter().map(|b| b.wrapping_add(1)).collect()
}

#[derive(Clone)]
pub struct Reader<R: Read> {
    inner: R,
}

impl<R: Read> Reader<R> {
    pub fn new(inner: R) -> Self {
        Self { inner }
    }
}

impl<R: Read> Read for Reader<R> {
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        self.inner.read(buf).map(|result| {
            for byte in buf.iter_mut() {
                *byte = byte.wrapping_sub(1);
            }
            result
        })
    }
}

#[cfg(test)]
mod tests {
    use super::{encrypt, Reader};
    use std::io::copy;

    fn decrypt(s: &[u8]) -> Vec<u8> {
        let mut reader = Reader::new(s);
        let mut vec = Vec::with_capacity(s.len());
        copy(&mut reader, &mut vec).unwrap();
        vec
    }

    #[test]
    fn decryption_works() {
        assert_eq!(
            decrypt(b"=@ynm!wfstjpo>#2/1#!fodpejoh>#VUG.9#@?=GMBSNOFU!Wfstjpo>#11:44f#?=GMBSNEBUB!GmbsnJE>#111111#?"),
            b"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<FLARMNET Version=\"00933e\">\n<FLARMDATA FlarmID=\"000000\">\n"
        );
    }

    #[test]
    fn encryption_works() {
        assert_eq!(
            encrypt(b"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<FLARMNET Version=\"00933e\">\n<FLARMDATA FlarmID=\"000000\">\n"),
            b"=@ynm!wfstjpo>#2/1#!fodpejoh>#VUG.9#@?=GMBSNOFU!Wfstjpo>#11:44f#?=GMBSNEBUB!GmbsnJE>#111111#?"
        );
    }
}
