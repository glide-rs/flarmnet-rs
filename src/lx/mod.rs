//! Decoder/Encoder for LXNav/Naviter file format.
//!
//! The file format is based on XML and "encrypted" by shifting each byte
//! by one.
//!
//! After decryption the XML structure looks like this:
//! ```xml
//! <?xml version="1.0" encoding="UTF-8" ?>
//! <FLARMNET Version="012345">
//!   <FLARMDATA FlarmID="000001">
//!     <NAME>John Doe</NAME>
//!     <AIRFIELD>EDKA</AIRFIELD>
//!     <TYPE>ASG 29</TYPE>
//!     <REG>D-KESH</REG>
//!     <COMPID>AS</COMPID>
//!     <FREQUENCY>123.500</FREQUENCY>
//!   </FLARMDATA>
//! </FLARMNET>
//! ```
//!
//! The [decode_file] function can be used to decode FlarmNet files in
//! LXNav/Naviter file format. The [encode_file] function can be used to write
//! such files.

pub mod cipher;
mod decode;
mod encode;

pub use decode::*;
pub use encode::*;
