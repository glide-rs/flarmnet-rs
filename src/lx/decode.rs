use crate::lx::cipher;
use crate::Record;
use minidom::{Element, NSChoice};
use std::string::FromUtf8Error;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum DecodeError {
    #[error(transparent)]
    Xml(#[from] minidom::Error),
    #[error(transparent)]
    Utf8(#[from] FromUtf8Error),
    #[error("missing XML element: {0}")]
    MissingElement(String),
    #[error("missing file version")]
    MissingVersion,
    #[error("invalid file version: {0}")]
    InvalidVersion(String),
    #[error("missing FLARM id")]
    MissingFlarmId,
    #[error("invalid FLARM id: {0}")]
    InvalidFlarmId(String),
}

#[derive(Debug)]
pub struct DecodedFile {
    pub version: u32,
    pub records: Vec<Result<Record, DecodeError>>,
}

/// Decodes a FlarmNet file in LX format.
///
/// # Examples
///
/// ```
/// # use flarmnet::Record;
/// let file = r#"=@ynm!wfstjpo>#2/1#!fodpejoh>#VUG.9#@?\u{b}=GMBSNOFU!Wfstjpo>#117gc1#?\u{b}=GMBSNEBUB!GmbsnJE>#111111#?\u{b}
/// =OBNF?NÄ½mmfs=0OBNF?\u{b}
/// =BJSGJFME?E.3299=0BJSGJFME?\u{b}
/// =UZQF?BTL.24=0UZQF?\u{b}
/// =SFH?E.3299=0SFH?\u{b}
/// =DPNQJE?=0DPNQJE?\u{b}
/// =GSFRVFODZ?234/261=0GSFRVFODZ?\u{b=0GMBSNEBUB?\u{b=GMBSNEBUB!GmbsnJE>#111112#?\u{b}
/// =OBNF?=0OBNF?\u{b}
/// =BJSGJFME?111111=0BJSGJFME?\u{b}
/// =UZQF?Qbsbhmjefs=0UZQF?\u{b}
/// =SFH?111111=0SFH?\u{b}
/// =DPNQJE?=0DPNQJE?\u{b}
/// =GSFRVFODZ?=0GSFRVFODZ?\u{b=0GMBSNEBUB?\u{b=GMBSNEBUB!GmbsnJE>#11111G#?\u{b}
/// =OBNF?=0OBNF?\u{b}
/// =BJSGJFME?E.:638=0BJSGJFME?\u{b}
/// =UZQF?BTX!38=0UZQF?\u{b}
/// =SFH?E.:638=0SFH?\u{b}
/// =DPNQJE?Y38=0DPNQJE?\u{b}
/// =GSFRVFODZ?=0GSFRVFODZ?\u{b}=0GMBSNEBUB?=0GMBSNOFU?
/// "#;
///
/// let result = flarmnet::lx::decode_file(file).unwrap();
/// assert_eq!(result.version, 28592);
/// assert_eq!(result.records.iter().filter(|it| it.is_ok()).count(), 3);
/// ```
pub fn decode_file(file: &str) -> Result<DecodedFile, DecodeError> {
    let decrypted = cipher::decrypt(file)?;

    let root: Element = decrypted.parse()?;
    if root.name() != "FLARMNET" {
        return Err(DecodeError::MissingElement("FLARMNET".to_string()));
    }

    let version = root.attr("Version").ok_or(DecodeError::MissingVersion)?;
    let version = u32::from_str_radix(version, 16)
        .map_err(|_| DecodeError::InvalidVersion(version.to_string()))?;

    let records = root
        .children()
        .filter(|child| child.name() == "FLARMDATA")
        .map(|child| convert(child))
        .collect();

    Ok(DecodedFile { version, records })
}

/// Converts a `minidom::Element` to a `flarmnet::Record`.
///
/// Expected structure:
///
/// ```xml
/// <FLARMDATA FlarmID="000001">
///   <NAME></NAME>
///   <AIRFIELD>000000</AIRFIELD>
///   <TYPE>Paraglider</TYPE>
///   <REG>000000</REG>
///   <COMPID></COMPID>
///   <FREQUENCY></FREQUENCY>
/// </FLARMDATA>
/// ```
fn convert(element: &Element) -> Result<Record, DecodeError> {
    let flarm_id = element
        .attr("FlarmID")
        .ok_or(DecodeError::MissingFlarmId)?
        .to_string();

    if u32::from_str_radix(&flarm_id, 16).is_err() {
        return Err(DecodeError::InvalidFlarmId(flarm_id));
    }

    let pilot_name = element
        .get_child("NAME", NSChoice::Any)
        .map(|e| e.text())
        .unwrap_or_default();

    let airfield = element
        .get_child("AIRFIELD", NSChoice::Any)
        .map(|e| e.text())
        .unwrap_or_default();

    let plane_type = element
        .get_child("TYPE", NSChoice::Any)
        .map(|e| e.text())
        .unwrap_or_default();

    let registration = element
        .get_child("REG", NSChoice::Any)
        .map(|e| e.text())
        .unwrap_or_default();

    let call_sign = element
        .get_child("COMPID", NSChoice::Any)
        .map(|e| e.text())
        .unwrap_or_default();

    let frequency = element
        .get_child("FREQUENCY", NSChoice::Any)
        .map(|e| e.text())
        .unwrap_or_default();

    Ok(Record {
        flarm_id,
        pilot_name,
        airfield,
        plane_type,
        registration,
        call_sign,
        frequency,
    })
}

#[cfg(test)]
mod tests {
    use crate::lx::cipher::encrypt;
    use crate::lx::decode::{convert, decode_file};
    use insta::assert_debug_snapshot;
    use minidom::Element;

    #[test]
    fn decoding_fails_for_empty_file() {
        let file = "";
        assert_debug_snapshot!(decode_file(file).unwrap_err(), @r###"
        Xml(
            EndOfDocument,
        )
        "###);
    }

    #[test]
    fn decoding_fails_for_invalid_file() {
        let file = "foo";
        assert_debug_snapshot!(decode_file(file).unwrap_err(), @r###"
        Xml(
            EndOfDocument,
        )
        "###);
    }

    #[test]
    fn decoding_fails_for_missing_root_element() {
        let file = encrypt(
            r#"<?xml version="1.0" encoding="UTF-8"?>
                <FOO>
                </FOO>"#,
        )
        .unwrap();
        assert_debug_snapshot!(decode_file(&file).unwrap_err(), @r###"
        MissingElement(
            "FLARMNET",
        )
        "###);
    }

    #[test]
    fn decoding_fails_for_missing_file_version() {
        let file = encrypt(
            r#"<?xml version="1.0" encoding="UTF-8"?>
                <FLARMNET>
                </FLARMNET>"#,
        )
        .unwrap();
        assert_debug_snapshot!(decode_file(&file).unwrap_err(), @"MissingVersion");
    }

    #[test]
    fn decoding_fails_for_invalid_file_version() {
        let file = encrypt(
            r#"<?xml version="1.0" encoding="UTF-8"?>
                <FLARMNET Version="foo">
                </FLARMNET>"#,
        )
        .unwrap();
        assert_debug_snapshot!(decode_file(&file).unwrap_err(), @r###"
        InvalidVersion(
            "foo",
        )
        "###);
    }

    #[test]
    fn converting_works() {
        let element: Element = r#"
            <FLARMDATA FlarmID="c0ffee">
              <NAME>John Doe</NAME>
              <AIRFIELD>EDKA</AIRFIELD>
              <TYPE>ASG 29</TYPE>
              <REG>D-KESH</REG>
              <COMPID>AS</COMPID>
              <FREQUENCY>123.500</FREQUENCY>
            </FLARMDATA>
        "#
        .parse()
        .unwrap();

        assert_debug_snapshot!(convert(&element).unwrap(), @r###"
        Record {
            flarm_id: "c0ffee",
            pilot_name: "John Doe",
            airfield: "EDKA",
            plane_type: "ASG 29",
            registration: "D-KESH",
            call_sign: "AS",
            frequency: "123.500",
        }
        "###);
    }

    #[test]
    fn converting_fails_for_missing_flarm_id() {
        let element: Element = r#"
            <FLARMDATA>
              <NAME></NAME>
              <AIRFIELD>000000</AIRFIELD>
              <TYPE>Paraglider</TYPE>
              <REG>000000</REG>
              <COMPID></COMPID>
              <FREQUENCY></FREQUENCY>
            </FLARMDATA>
        "#
        .parse()
        .unwrap();

        assert_debug_snapshot!(convert(&element).unwrap_err(), @"MissingFlarmId");
    }

    #[test]
    fn converting_fails_for_invalid_flarm_id() {
        let element: Element = r#"
            <FLARMDATA FlarmID="foo">
              <NAME></NAME>
              <AIRFIELD>000000</AIRFIELD>
              <TYPE>Paraglider</TYPE>
              <REG>000000</REG>
              <COMPID></COMPID>
              <FREQUENCY></FREQUENCY>
            </FLARMDATA>
        "#
        .parse()
        .unwrap();

        assert_debug_snapshot!(convert(&element).unwrap_err(), @r###"
        InvalidFlarmId(
            "foo",
        )
        "###);
    }
}
