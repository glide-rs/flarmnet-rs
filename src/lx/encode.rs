use crate::lx::cipher::encrypt;
use crate::File;
use minidom::quick_xml;
use minidom::quick_xml::events::{BytesDecl, BytesEnd, BytesStart, BytesText, Event};
use minidom::quick_xml::Writer;
use std::io::Cursor;
use thiserror::Error;

#[derive(Error, Debug)]
pub enum EncodeError {
    #[error(transparent)]
    Xml(#[from] quick_xml::Error),
}

/// Encodes a FlarmNet file in LX format.
///
/// # Examples
///
/// ```
/// # use flarmnet::Record;
/// let file = flarmnet::File {
///     version: 123,
///     records: vec![
///         Record {
///             flarm_id: "3EE3C7".to_string(),
///             pilot_name: "Tobias Bieniek".to_string(),
///             airfield: "EDKA".to_string(),
///             plane_type: "LS6a".to_string(),
///             registration: "D-0816".to_string(),
///             call_sign: "SG".to_string(),
///             frequency: "130.530".to_string(),
///         }
///     ]
/// };
///
/// let result = flarmnet::lx::encode_file(&file);
/// assert!(result.is_ok());
/// ```
pub fn encode_file(file: &File) -> Result<Vec<u8>, EncodeError> {
    let mut writer = Writer::new_with_indent(Cursor::new(Vec::new()), 9u8, 0);

    writer.write_event(Event::Decl(BytesDecl::new(b"1.0", Some(b"UTF-8"), None)))?;

    let version = format!("{:06x?}", file.version);
    writer.write_event(Event::Start(
        BytesStart::borrowed_name(b"FLARMNET")
            .with_attributes(vec![("Version".as_bytes(), version.as_bytes())]),
    ))?;
    for record in &file.records {
        writer.write_event(Event::Start(
            BytesStart::borrowed_name(b"FLARMDATA")
                .with_attributes(vec![("FlarmID".as_bytes(), record.flarm_id.as_bytes())]),
        ))?;

        writer.write_event(Event::Text(BytesText::from_plain_str("\n\t")))?;
        writer.write_event(Event::Start(BytesStart::borrowed_name(b"NAME")))?;
        writer.write_event(Event::Text(BytesText::from_plain_str(&record.pilot_name)))?;
        writer.write_event(Event::End(BytesEnd::borrowed(b"NAME")))?;

        writer.write_event(Event::Text(BytesText::from_plain_str("\n\t")))?;
        writer.write_event(Event::Start(BytesStart::borrowed_name(b"AIRFIELD")))?;
        writer.write_event(Event::Text(BytesText::from_plain_str(&record.airfield)))?;
        writer.write_event(Event::End(BytesEnd::borrowed(b"AIRFIELD")))?;

        writer.write_event(Event::Text(BytesText::from_plain_str("\n\t")))?;
        writer.write_event(Event::Start(BytesStart::borrowed_name(b"TYPE")))?;
        writer.write_event(Event::Text(BytesText::from_plain_str(&record.plane_type)))?;
        writer.write_event(Event::End(BytesEnd::borrowed(b"TYPE")))?;

        writer.write_event(Event::Text(BytesText::from_plain_str("\n\t")))?;
        writer.write_event(Event::Start(BytesStart::borrowed_name(b"REG")))?;
        writer.write_event(Event::Text(BytesText::from_plain_str(&record.registration)))?;
        writer.write_event(Event::End(BytesEnd::borrowed(b"REG")))?;

        writer.write_event(Event::Text(BytesText::from_plain_str("\n\t")))?;
        writer.write_event(Event::Start(BytesStart::borrowed_name(b"COMPID")))?;
        writer.write_event(Event::Text(BytesText::from_plain_str(&record.call_sign)))?;
        writer.write_event(Event::End(BytesEnd::borrowed(b"COMPID")))?;

        writer.write_event(Event::Text(BytesText::from_plain_str("\n\t")))?;
        writer.write_event(Event::Start(BytesStart::borrowed_name(b"FREQUENCY")))?;
        writer.write_event(Event::Text(BytesText::from_plain_str(&record.frequency)))?;
        writer.write_event(Event::End(BytesEnd::borrowed(b"FREQUENCY")))?;

        writer.write_event(Event::End(BytesEnd::borrowed(b"FLARMDATA")))?;
    }
    writer.write_event(Event::End(BytesEnd::borrowed(b"FLARMNET")))?;

    let xml = writer.into_inner().into_inner();

    Ok(encrypt(&xml))
}
