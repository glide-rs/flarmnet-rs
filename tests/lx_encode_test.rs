use flarmnet::lx::{decode_file, encode_file};
use insta::assert_snapshot;

#[test]
fn it_works() {
    let fixture = include_bytes!("fixtures/lx.fln");
    let decoded = decode_file(fixture).unwrap();
    let file = flarmnet::File {
        version: decoded.version,
        records: decoded
            .records
            .into_iter()
            .filter_map(|it| it.ok())
            .collect(),
    };
    let encoded = encode_file(&file).unwrap();
    assert_snapshot!(String::from_utf8_lossy(&encoded));
}
