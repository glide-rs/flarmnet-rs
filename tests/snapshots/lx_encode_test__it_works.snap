---
source: tests/lx_encode_test.rs
expression: "String::from_utf8_lossy(&encoded)"
---
=@ynm!wfstjpo>#2/1#!fodpejoh>#VUG.9#@?=GMBSNOFU!Wfstjpo>#117gc1#?=GMBSNEBUB!GmbsnJE>#111111#?
=OBNF?NÄ½mmfs=0OBNF?
=BJSGJFME?E.3299=0BJSGJFME?
=UZQF?BTL.24=0UZQF?
=SFH?E.3299=0SFH?
=DPNQJE?=0DPNQJE?
=GSFRVFODZ?234/261=0GSFRVFODZ?=0GMBSNEBUB?=GMBSNEBUB!GmbsnJE>#111112#?
=OBNF?=0OBNF?
=BJSGJFME?111111=0BJSGJFME?
=UZQF?Qbsbhmjefs=0UZQF?
=SFH?111111=0SFH?
=DPNQJE?=0DPNQJE?
=GSFRVFODZ?=0GSFRVFODZ?=0GMBSNEBUB?=GMBSNEBUB!GmbsnJE>#11111G#?
=OBNF?=0OBNF?
=BJSGJFME?E.:638=0BJSGJFME?
=UZQF?BTX!38=0UZQF?
=SFH?E.:638=0SFH?
=DPNQJE?Y38=0DPNQJE?
=GSFRVFODZ?=0GSFRVFODZ?=0GMBSNEBUB?=0GMBSNOFU?
